File: src/main/java/services/GeminiRecommendationService.java
package services;

import io.github.cdimascio.dotenv.Dotenv;
import models.Activite;

import java.net.URI;
import java.net.http.*;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.*;
import java.util.stream.Collectors;

public class GeminiRecommendationService {

    private static final boolean DEBUG = true;

    private static final Dotenv dotenv = Dotenv.load();
    private static final String API_KEY = dotenv.get("GEMINI_API_KEY");

    private static final String ENDPOINT =
            "https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash:generateContent?key=";

    // ✅ Timeout pour éviter attente longue
    private static final Duration TIMEOUT = Duration.ofSeconds(20);

    public List<Integer> rankActivityIdsMax3(List<Activite> candidates, String userProfileText) throws Exception {
        if (API_KEY == null || API_KEY.isBlank()) {
            if (DEBUG) System.out.println("[Gemini] API_KEY manquante");
            return List.of();
        }
        if (candidates == null || candidates.isEmpty()) return List.of();

        // ✅ Limite candidats (rapide)
        List<Activite> list = candidates.stream().limit(18).toList();

        // ✅ payload léger: description courte
        String activitiesJson = list.stream()
                .map(a -> String.format(
                        "{\"id\":%d,\"name\":%s,\"type\":%s,\"price\":%.2f,\"rating\":%.1f,\"desc\":%s}",
                        a.getId(),
                        js(safe(a.getNom())),
                        js(safe(a.getTypeActivite())),
                        a.getPrix(),
                        a.getNoteMoyenne(),
                        js(shortText(safe(a.getDescription()), 80))
                ))
                .collect(Collectors.joining(",", "[", "]"));

        String prompt = """
        Tu es un moteur de recommandation d’activités de voyage.
        Choisis et classe entre 0 et 3 activités maximum adaptées au profil utilisateur.
        Réponds UNIQUEMENT avec du JSON minifié, sans markdown, sans backticks.

        Format EXACT attendu :
        {"ranked_ids":[...]}  // 0..3 ids existants dans la liste

        Profil utilisateur:
        %s

        Activités candidates (JSON):
        %s

        Règles:
        - Respecte budgetMin/budgetMax si possible
        - Match centresInteret/typesVoyage avec name/type/desc
        - Si aucune activité ne correspond, retourne {"ranked_ids":[]}
        - Ne renvoie jamais plus de 3 IDs
        """.formatted(userProfileText, activitiesJson);

        String body = """
        {
          "contents": [
            { "parts": [ { "text": %s } ] }
          ]
        }
        """.formatted(js(prompt));

        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(ENDPOINT + API_KEY))
                .timeout(TIMEOUT)
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(body, StandardCharsets.UTF_8))
                .build();

        HttpClient client = HttpClient.newHttpClient();

        HttpResponse<String> response = null;

        for (int attempt = 1; attempt <= 2; attempt++) {
            try {
                response = client.send(request, HttpResponse.BodyHandlers.ofString());
                break;
            } catch (Exception ex) {
                if (DEBUG) System.out.println("[Gemini] attempt " + attempt + " failed: " + ex.getMessage());
                if (attempt == 2) return List.of();
                Thread.sleep(500);
            }
        }

        if (DEBUG) {
            System.out.println("\n===== GEMINI RAW RESPONSE (HTTP " + response.statusCode() + ") =====");
            System.out.println(response.body());
            System.out.println("===============================================================\n");
        }

        if (response.statusCode() != 200) return List.of();

        String modelText = extractModelText(response.body());
        if (modelText == null) return List.of();

        if (DEBUG) {
            System.out.println("===== GEMINI MODEL TEXT =====");
            System.out.println(modelText);
            System.out.println("=============================\n");
        }

        modelText = modelText.replace("```json", "").replace("```", "").trim();

        List<Integer> ids = extractRankedIds(modelText);
        if (ids.size() > 3) ids = ids.subList(0, 3);

        if (DEBUG) {
            System.out.println("===== EXTRACTED ranked_ids =====");
            System.out.println(ids);
            System.out.println("================================\n");
        }

        return ids;
    }
    // ---- parse candidates[0].content.parts[0].text ----
    private String extractModelText(String json) {
        int i = json.indexOf("\"text\"");
        if (i < 0) return null;
        int colon = json.indexOf(':', i);
        if (colon < 0) return null;
        int start = json.indexOf('"', colon + 1);
        if (start < 0) return null;

        StringBuilder sb = new StringBuilder();
        boolean escape = false;

        for (int k = start + 1; k < json.length(); k++) {
            char c = json.charAt(k);

            if (escape) {
                if (c == 'n') sb.append('\n');
                else if (c == 't') sb.append('\t');
                else sb.append(c);
                escape = false;
            } else {
                if (c == '\\') escape = true;
                else if (c == '"') break;
                else sb.append(c);
            }
        }
        return sb.toString();
    }

    private List<Integer> extractRankedIds(String json) {
        int key = json.indexOf("\"ranked_ids\"");
        if (key < 0) return List.of();

        int start = json.indexOf('[', key);
        int end = json.indexOf(']', start);
        if (start < 0 || end < 0) return List.of();

        String inside = json.substring(start + 1, end).trim();
        if (inside.isEmpty()) return List.of();

        List<Integer> ids = new ArrayList<>();
        for (String p : inside.split(",")) {
            try { ids.add(Integer.parseInt(p.trim())); } catch (Exception ignored) {}
        }
        return ids;
    }

    private static String safe(String s) { return s == null ? "" : s; }

    private static String shortText(String s, int max) {
        if (s == null) return "";
        s = s.replace("\n", " ").replace("\r", " ").trim();
        if (s.length() <= max) return s;
        return s.substring(0, max);
    }

    private static String js(String s) {
        return "\"" + s.replace("\\","\\\\").replace("\"","\\\"") + "\"";
    }
}

File: src/Service/GeminiRecommendationService.php
<?php

namespace App\Service;

use App\Entity\Activite;
use Symfony\Contracts\Cache\CacheInterface;
use Symfony\Contracts\Cache\ItemInterface;
use Symfony\Contracts\HttpClient\HttpClientInterface;

class GeminiRecommendationService
{
    private const ENDPOINT = 'https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash:generateContent?key=';
    private const DEBUG = true;

    public function __construct(
        private HttpClientInterface $httpClient,
        private string $geminiApiKey,
        private CacheInterface $cache
    ) {
    }

    /**
     * Cached version:
     * Gemini is called again only if user preferences or activities changed.
     *
     * @param Activite[] $candidates
     * @return int[]
     */
    public function rankActivityIdsMax3Cached(
        int $userId,
        array $candidates,
        string $userProfileText
    ): array {
        if (empty($candidates) || empty($userProfileText) || empty($this->geminiApiKey)) {
            return [];
        }

        $cacheKey = $this->buildCacheKey($userId, $userProfileText, $candidates);

        return $this->cache->get($cacheKey, function (ItemInterface $item) use ($candidates, $userProfileText) {
            // Long lifetime, but real refresh happens when cache key changes
            $item->expiresAfter(60 * 60 * 24 * 30); // 30 days

            return $this->rankActivityIdsMax3($candidates, $userProfileText);
        });
    }

    /**
     * Real Gemini API call
     *
     * @param Activite[] $candidates
     * @return int[]
     */
    public function rankActivityIdsMax3(array $candidates, string $userProfileText): array
    {
        if (empty($this->geminiApiKey)) {
            return [];
        }

        if (empty($candidates)) {
            return [];
        }

        $list = array_slice($candidates, 0, 18);

        $activities = array_map(function (Activite $a) {
            return [
                'id' => $a->getId(),
                'name' => $this->safe($a->getNom()),
                'type' => $this->safe($a->getTypeActivite()),
                'price' => (float) $a->getPrix(),
                'rating' => (float) $a->getNoteMoyenne(),
                'desc' => $this->shortText($this->safe($a->getDescription()), 80),
            ];
        }, $list);

        $prompt = sprintf(
            "Tu es un moteur de recommandation d’activités de voyage.
Choisis et classe entre 0 et 3 activités maximum adaptées au profil utilisateur.
Réponds UNIQUEMENT avec du JSON minifié, sans markdown, sans backticks.

Format EXACT attendu :
{\"ranked_ids\":[...]}

Profil utilisateur:
%s

Activités candidates (JSON):
%s

Règles:
- Respecte budgetMin/budgetMax si possible
- Match centresInteret/typesVoyage avec name/type/desc
- Si aucune activité ne correspond, retourne {\"ranked_ids\":[]}
- Ne renvoie jamais plus de 3 IDs
- N’utilise que des IDs existants dans la liste",
            $userProfileText,
            json_encode($activities, JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES)
        );

        $body = [
            'contents' => [
                [
                    'parts' => [
                        ['text' => $prompt]
                    ]
                ]
            ]
        ];

        try {
            $response = $this->httpClient->request('POST', self::ENDPOINT . $this->geminiApiKey, [
                'headers' => [
                    'Content-Type' => 'application/json',
                ],
                'json' => $body,
                'timeout' => 20,
            ]);

            $statusCode = $response->getStatusCode();
            $raw = $response->getContent(false);

            if (self::DEBUG) {
                dump([
                    'gemini_status' => $statusCode,
                    'gemini_raw' => $raw,
                ]);
            }

            if ($statusCode !== 200) {
                return [];
            }

            $data = json_decode($raw, true);

            $modelText = $data['candidates'][0]['content']['parts'][0]['text'] ?? null;

            if (!$modelText) {
                return [];
            }

            $modelText = trim(str_replace(['```json', '```'], '', $modelText));

            preg_match('/"ranked_ids"\s*:\s*\[(.*?)\]/s', $modelText, $matches);

            if (!isset($matches[1])) {
                return [];
            }

            $inside = trim($matches[1]);

            if ($inside === '') {
                return [];
            }

            $ids = array_filter(array_map(function ($value) {
                $value = trim($value);
                return is_numeric($value) ? (int) $value : null;
            }, explode(',', $inside)));

            $ids = array_values(array_unique($ids));

            return array_slice($ids, 0, 3);
        } catch (\Throwable $e) {
            if (self::DEBUG) {
                dump($e->getMessage());
            }

            return [];
        }
    }

    private function buildPreferencesFingerprint(string $userProfileText): string
    {
        return md5($userProfileText);
    }

    /**
     * @param Activite[] $candidates
     */
    private function buildActivitiesFingerprint(array $candidates): string
    {
        $data = array_map(function (Activite $a) {
            return [
                'id' => $a->getId(),
                'nom' => (string) $a->getNom(),
                'type' => (string) $a->getTypeActivite(),
                'prix' => (float) $a->getPrix(),
                'rating' => (float) $a->getNoteMoyenne(),
                'description' => (string) $a->getDescription(),
                'status' => method_exists($a, 'getStatus') ? (string) $a->getStatus() : null,
                'maxPlaces' => method_exists($a, 'getMaxPlaces') ? (int) $a->getMaxPlaces() : null,
            ];
        }, $candidates);

        return md5(json_encode($data, JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES));
    }

    /**
     * @param Activite[] $candidates
     */
    private function buildCacheKey(
        int $userId,
        string $userProfileText,
        array $candidates
    ): string {
        $preferencesFingerprint = $this->buildPreferencesFingerprint($userProfileText);
        $activitiesFingerprint = $this->buildActivitiesFingerprint($candidates);

        return 'gemini_reco_' . $userId . '_' . $preferencesFingerprint . '_' . $activitiesFingerprint;
    }

    private function safe(?string $value): string
    {
        return $value ?? '';
    }

    private function shortText(string $text, int $max): string
    {
        $text = trim(str_replace(["\n", "\r"], ' ', $text));

        if (mb_strlen($text) <= $max) {
            return $text;
        }

        return mb_substr($text, 0, $max);
    }
}